src/Uno.UWP/ApplicationModel/DataTransfer/DataPackage.cs

```diff
@@ -1,73 +1,97 @@
-﻿#if !NET461
+﻿#nullable enable
+
+#if !NET461
 using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Linq;
+using System.Threading;
 using System.Threading.Tasks;
+using Windows.Foundation;
+using Windows.Storage;
+using Windows.Storage.Streams;
+using Uno;
+using Uno.Extensions;
+using Uno.Logging;
 
 namespace Windows.ApplicationModel.DataTransfer
 {
 	public partial class DataPackage
 	{
-		internal string Text { get; private set; }
-
-		internal string Html { get; private set; }
+		public event TypedEventHandler<DataPackage, OperationCompletedEventArgs>? OperationCompleted;
 
-		internal Uri Uri { get; private set; }
+		private ImmutableDictionary<string, object> _data = ImmutableDictionary<string, object>.Empty;
 
 		public DataPackageOperation RequestedOperation { get; set; }
 
-		public void SetText(string value)
+		public IDictionary<string, RandomAccessStreamReference> ResourceMap { get; } = new Dictionary<string, RandomAccessStreamReference>();
+
+		public void SetData(string formatId, object value)
 		{
-			if (value == null)
-			{
-				throw new ArgumentNullException("Text can't be null");
-			}
+			ImmutableInterlocked.Update(ref _data, SetDataCore, (formatId, value));
 
-			Text = value;
+			ImmutableDictionary<string, object> SetDataCore(ImmutableDictionary<string, object> current, (string id, object v) arg)
+				=> current.SetItem(arg.id, arg.v);
 		}
 
-		public void SetUri(Uri value)
+		internal void SetDataProvider(string formatId, FuncAsync<object> delayRenderer)
 		{
-			if (value == null)
+			SetData(formatId, new DataProviderHandler(SetDataCore));
+
+			async void SetDataCore(DataProviderRequest request)
 			{
-				throw new ArgumentNullException("Cannot set DataPackage.Uri to null");
+				var deferral = request.GetDeferral();
+				try
+				{
+					await delayRenderer(request.CancellationToken);
```

Comment:
This will not work. The delayRenderer data is going nowhere. To fix this in my branch I added:

```csharp
var data = await delayRenderer(request.CancellationToken);
request.SetData(data);
```

Reply:
Whoups ... good point! I applied your fix.